core/src/main/scala/com/evolutiongaming/catshelper/Log.scala

```diff
@@ -44,30 +44,66 @@ object Log {
   object Mdc {
 
     private object Empty extends Mdc
-    private final case class Context(values: NonEmptyMap[String, String]) extends Mdc {
-      override def toString: String = s"MDC(${values.toSortedMap.mkString(", ")})"
-    }
+    private final case class Context(getValues: () => Map[String, String]) extends Mdc {
+      lazy val values: SortedMap[String, String] = SortedMap(getValues().toSeq: _*)
 
-    val empty: Mdc = Empty
+      override def toString: String = s"MDC(${values.mkString(", ")})"
 
-    def apply(head: (String, String), tail: (String, String)*): Mdc = Context(NonEmptyMap.of(head, tail: _*))
+      override def hashCode(): Int = values.hashCode()
 
-    def fromSeq(seq: Seq[(String, String)]): Mdc =
-      NonEmptyMap.fromMap(SortedMap(seq: _*)).fold(empty){ nem => Context(nem) }
+      override def equals(obj: Any): Boolean = obj match {
+        case that: Context => this.values.equals(that.values)
+        case _ => false
+      }
+    }
 
-    def fromMap(map: Map[String, String]): Mdc = fromSeq(map.toSeq)
+    val empty: Mdc = Empty
+
+    type Record = (String, String)
+
+    @deprecated("Use Mdc.Lazy.apply. If it's not enough - Mdc.Lazy.fromMap", "3.9.0")
+    def apply(head: Record, tail: Record*): Mdc = Lazy.fromMap( (head +: tail).toMap)
```

Comment:
Please do not deprecate those factory methods, in many places its fine to have eager construction

Reply:
Something like Mdc("key" -> parseSomethingFromJson(...)) can be heavy and would be nice to be lazy

Reply:
yes, but its up to user to choose eager/lazy to use

Reply:
When do you think the eager implementation is preferred and a user wants to choose it instead of the lazy one?

Reply:
1. when MDC already pre-calculated 
2. when MDC is compact and all its values already available

I would say that almost everywhere I prefer eager. Lazy one is corner case in the apps with wide context

Reply:
Discussed with @t3hnar and agreed to leave all these three methods as deprecated and implement another ones. Users will have to decide which one will be used